File: backend/libs/nova_contracts/src/nova_contracts/realtime.py
"""WebSocket messages (D57): mirror `frontend/packages/contracts/src/realtime.ts`."""

from typing import Annotated, Literal

from pydantic import Field

from nova_contracts.common import Contract
from nova_contracts.data_job import DataJob


class RealtimeHello(Contract):
    """First message after the socket opens."""

    type: Literal["hello"]


class RealtimePing(Contract):
    """Sent every 25 s; the client answers `{"type": "pong"}`."""

    type: Literal["ping"]


class DataJobUpdated(Contract):
    """A data job was created or changed (any type, D57)."""

    type: Literal["data_job.updated"]
    data: DataJob


RealtimeMessage = Annotated[
    RealtimeHello | RealtimePing | DataJobUpdated, Field(discriminator="type")
]

File: frontend/packages/contracts/src/realtime.ts
import { z } from "zod";
import { DataJobSchema } from "./dataJob";

/** WebSocket `/api/v1/ws` messages from NOVA Core (D57). */
export const RealtimeMessageSchema = z.discriminatedUnion("type", [
  /** First message after the socket opens. */
  z.strictObject({ type: z.literal("hello") }),
  /** Every 25 s; answer `{ type: "pong" }`. */
  z.strictObject({ type: z.literal("ping") }),
  /** A data job was created or changed (any type). */
  z.strictObject({ type: z.literal("data_job.updated"), data: DataJobSchema }),
]);
export type RealtimeMessage = z.infer<typeof RealtimeMessageSchema>;
